graphkit/network.py

```diff
@@ -214,9 +215,86 @@ def compute(self, outputs, named_inputs):
 
         # assert that network has been compiled
         assert self.steps, "network must be compiled before calling compute."
-        assert isinstance(outputs, (list, tuple)) or outputs == None,\
+        assert isinstance(outputs, (list, tuple)) or outputs is None,\
             "The outputs argument must be a list"
 
+
+        # choose a method of execution
+        if method == "parallel":
+            return self._compute_thread_pool_barrier_method(named_inputs,
+                                                            outputs)
+        else:
+            return self._compute_sequential_method(named_inputs,
+                                                   outputs)
+
+
+    def _compute_thread_pool_barrier_method(self, named_inputs, outputs,
+                                            thread_pool_size=10):
+        """
+        This method runs the graph using a parallel pool of thread executors.
+        You may achieve lower total latency if your graph is sufficiently
+        sub divided into operations using this method.
+        """
+        from multiprocessing.dummy import Pool
+
+        # if we have not already created a thread_pool, create one
+        if not hasattr(self, "_pool"):
+            self._pool = Pool(thread_pool_size)
+        pool = self._pool
+
+        cache = {}
+        cache.update(named_inputs)
+        necessary_nodes = self._find_necessary_steps(outputs, named_inputs)
+
+        # this keeps track of all nodes that have already executed
+        has_executed = set()
+
+        # with each loop iteration, we determine a set of operations that can be
+        # scheduled, then schedule them onto a thread pool, then collect their
+        # results onto a memory cache for use upon the next iteration.
+        while True:
+
+            # the upnext list contains a list of operations for scheduling
+            # in the current round of scheduling
+            upnext = []
+            for node in necessary_nodes:
+                # only delete if all successors for the data node have been executed
+                if isinstance(node, DeleteInstruction):
+                    if ready_to_delete_data_node(node,
+                                                 has_executed,
+                                                 self.graph):
+                        if node in cache:
+                            cache.pop(node)
+
+                # continue if this node is anything but an operation node
+                if not isinstance(node, Operation):
+                    continue
+
+                if ready_to_schedule_operation(node, has_executed, self.graph) \
+                        and node not in has_executed:
+                    upnext.append(node)
+
+
+            # stop if no nodes left to schedule, exit out of the loop
+            if len(upnext) == 0:
+                break
+
+            done_iterator = pool.imap_unordered(
```

Comment:
Do I understand correctly, that this is layerwise parallelism? as in, if at a certain depth in the graph, we have a pool of 5 workers and 1 worker takes a long time, then the others have to wait?

Reply:
Yes, your description is correct. The current scheduling method for the parallel thread pool is to  "schedule everything that is able to be scheduled within this current iteration" . In the same iteration, wait for all answers to come back, and then repeat the process. This may be some times inefficient when we have for example 1 task that takes very long and many tasks that take short amount of time. 

There is another way to do the scheduling which could potentially be more efficient but I have not figured out how to implement it yet. The idea would be to schedule operations that are "schedulable" as soon as there is a free worker slot.

Reply:
ok. thanks for clarifying :) you're right, no idea how that would work using threads from within python. the pattern i'd use other places would be to have the outermost loop iterate over threads. whenever there is a job available from the job queue, run it. idk whether you can address threads like that with `multiprocessing`.